README.txt
= ImageProc

* http://wiretap.rubyforge.org/image_proc

== SYNPOSIS

    require "image_proc"
    ImageProc.resize("/tmp/source_file.jpg", :width => 128, :height => 200, :fill => :true)

== DESCRIPTION:

A no-frills image resizer, with pluggable backends. No extra software required on OS X

== INSTALL:

* sudo gem install image_proc

== LICENSE:

(The MIT License)

Copyright (c) 2009-2011 Julik Tarkhanov <me@julik.nl>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 2.1.1 / 2011-10-24

* Fancify image filtering used with ImageProcConvert

=== 2.1.0 / 2011-07-19

* Check if the convert bin is available and try to intuit it's path

=== 2.0.0 / 2009-01-09

* Make all methods return the destination path only

=== 1.0.3 / 2009-01-09

* Ensure the destination path is always returned

=== 1.0.2 / 2009-01-09

* Allow nil keys in ImageProc.resize() options and just scrub them.

=== 1.0.1 / 2009-01-09

* Remove the old geometry string API

=== 0.1.1 / 2009-01-09

* Small updates to Hoe and other software rot prevention

=== 0.1.0 / 2009-01-09

* 1 major enhancement

  * Birthday!



geom_test.rb
require 'test/unit'
require 'fileutils'
require 'resize_test_helper'
require 'image_proc'

class TestGeometryFitting < Test::Unit::TestCase
  def setup
    @x = ImageProc.new
    class << @x
      public :fit_sizes
    end
  end
  
  def test_fit_width
    bounds = [1024, 500]
    assert_equal [50, 24], @x.fit_sizes(bounds, :width => 50)

    bounds = [500, 1024]
    assert_equal [50, 102], @x.fit_sizes(bounds, :width => 50)
  end
  
  def test_fit_width_should_not_modify_values_if_image_fits_already
    bounds = [100, 400]
    assert_equal bounds, @x.fit_sizes(bounds, :width => 100)
  end

  def test_fit_height_should_not_modify_values_if_image_fits_already
    bounds = [100, 400]
    assert_equal bounds, @x.fit_sizes(bounds, :height => 400)
  end
  
  def test_fit_height
    bounds = [1024, 500]
    assert_equal [102, 50], @x.fit_sizes(bounds, :height => 50)

    bounds = [500, 1024]
    assert_equal [24, 50], @x.fit_sizes(bounds, :height => 50)
  end
  
  def test_fit_both
    bounds = 1024, 500
    assert_equal [70, 34], @x.fit_sizes(bounds, :height => 70, :width => 70)

    bounds = 500, 1024
    assert_equal [34, 70], @x.fit_sizes(bounds, :height => 70, :width => 70)
    
    bounds = 780, 520
    assert_equal [120, 80], @x.fit_sizes(bounds, :height => 120, :width => 120)
  
  end
  
  def test_fit_groks_strings_too
    bounds = [1024, 500]
    assert_equal [70, 34], @x.fit_sizes(bounds, :height => "70", :width => "70")
  end
  
  def test_fit_rejects_nil_values
    bounds = [1024, 500]
    assert_equal [50, 24], @x.fit_sizes(bounds, :width => 50, :height => nil)
    assert_equal [102, 50], @x.fit_sizes(bounds, :height => 50, :width => nil)
  end

  def test_fit_nevere_produces_zeroes
    bounds = [1000, 20]
    assert_equal [20, 1], @x.fit_sizes(bounds, :width => 20, :height => 20)
  end
end

class TestGeometryFittingWithCrop < Test::Unit::TestCase
  def setup
    @x = ImageProc.new
  end
  
  def test_fit_suare_into_smaller_square_requires_no_cropping
    bounds = [100, 100]
    assert_equal [20, 20], @x.fit_sizes_with_crop(bounds, :width => 20, :height => 20)
  end

  def test_fit_landscape_into_square_will_slice_off_left_and_right
    bounds = [768, 575]
    assert_equal [27, 20], @x.fit_sizes_with_crop(bounds, :width => 20, :height => 20)

    bounds = [200, 40]
    assert_equal [100, 20], @x.fit_sizes_with_crop(bounds, :width => 20, :height => 20)
  end

  def test_fit_portrait_into_square_will_slice_off_top_and_bottom
    bounds = [576, 768]
    assert_equal [300, 400], @x.fit_sizes_with_crop(bounds, :width => 300, :height => 300)
  end

  def test_fit_portrait_into_portrait_will_resize
    bounds = [576, 768]
    assert_equal [200, 267], @x.fit_sizes_with_crop(bounds, :width => 200, :height => 210)
  end

  def test_fit_landscape_into_portrait_will_resize
    bounds = [768, 576]
    assert_equal [280, 210], @x.fit_sizes_with_crop(bounds, :width => 200, :height => 210)
  end
  
  def test_fit_with_crop_does_not_lie_when_small_floats_might_be_involved
   bounds = [20, 1000]
   assert_equal bounds, @x.fit_sizes_with_crop(bounds, :width => 20, :height => 20)
  end
  
  def test_fit_with_crop_should_not_overnudge
    bounds = [780, 520]
    assert_equal [375, 250], @x.fit_sizes_with_crop(bounds, :width => 260, :height => 250)
  end
end


test/test_geom.rb
require 'test/unit'
require 'fileutils'
require 'resize_test_helper'
require 'image_proc'

class TestGeometryFitting < Test::Unit::TestCase
  def setup
    @x = ImageProc.new
    class << @x
      public :fit_sizes
    end
  end
  
  def test_fit_width
    bounds = [1024, 500]
    assert_equal [50, 24], @x.fit_sizes(bounds, :width => 50)
    assert_equal [50, 24], @x.fit_sizes(bounds, :width => 50, :height => nil)

    bounds = [500, 1024]
    assert_equal [50, 102], @x.fit_sizes(bounds, :width => 50)
  end
  
  def test_fit_width_should_not_modify_values_if_image_fits_already
    bounds = [100, 400]
    assert_equal bounds, @x.fit_sizes(bounds, :width => 100)
  end

  def test_fit_height_should_not_modify_values_if_image_fits_already
    bounds = [100, 400]
    assert_equal bounds, @x.fit_sizes(bounds, :height => 400)
  end
  
  def test_fit_height
    bounds = [1024, 500]
    assert_equal [102, 50], @x.fit_sizes(bounds, :height => 50)
    assert_equal [102, 50], @x.fit_sizes(bounds, :height => 50, :width => nil)

    bounds = [500, 1024]
    assert_equal [24, 50], @x.fit_sizes(bounds, :height => 50)
  end
  
  def test_fit_both
    bounds = 1024, 500
    assert_equal [70, 34], @x.fit_sizes(bounds, :height => 70, :width => 70)

    bounds = 500, 1024
    assert_equal [34, 70], @x.fit_sizes(bounds, :height => 70, :width => 70)
    
    bounds = 780, 520
    assert_equal [120, 80], @x.fit_sizes(bounds, :height => 120, :width => 120)
  end
  
  def test_fit_groks_strings_too
    bounds = [1024, 500]
    assert_equal [70, 34], @x.fit_sizes(bounds, :height => "70", :width => "70")
  end
  
  def test_fit_rejects_nil_values
    bounds = [1024, 500]
    assert_equal [50, 24], @x.fit_sizes(bounds, :width => 50, :height => nil)
    assert_equal [102, 50], @x.fit_sizes(bounds, :height => 50, :width => nil)
  end

  def test_fit_nevere_produces_zeroes
    bounds = [1000, 20]
    assert_equal [20, 1], @x.fit_sizes(bounds, :width => 20, :height => 20)
  end
end

class TestGeometryFittingWithCrop < Test::Unit::TestCase
  def setup
    @x = ImageProc.new
  end
  
  def test_fit_suare_into_smaller_square_requires_no_cropping
    bounds = [100, 100]
    assert_equal [20, 20], @x.fit_sizes_with_crop(bounds, :width => 20, :height => 20)
  end

  def test_fit_landscape_into_square_will_slice_off_left_and_right
    bounds = [768, 575]
    assert_equal [27, 20], @x.fit_sizes_with_crop(bounds, :width => 20, :height => 20)

    bounds = [200, 40]
    assert_equal [100, 20], @x.fit_sizes_with_crop(bounds, :width => 20, :height => 20)
  end

  def test_fit_portrait_into_square_will_slice_off_top_and_bottom
    bounds = [576, 768]
    assert_equal [300, 400], @x.fit_sizes_with_crop(bounds, :width => 300, :height => 300)
  end

  def test_fit_portrait_into_portrait_will_resize
    bounds = [576, 768]
    assert_equal [200, 267], @x.fit_sizes_with_crop(bounds, :width => 200, :height => 210)
  end

  def test_fit_landscape_into_portrait_will_resize
    bounds = [768, 576]
    assert_equal [280, 210], @x.fit_sizes_with_crop(bounds, :width => 200, :height => 210)
  end
  
  def test_fit_with_crop_does_not_lie_when_small_floats_might_be_involved
   bounds = [20, 1000]
   assert_equal bounds, @x.fit_sizes_with_crop(bounds, :width => 20, :height => 20)
  end
  
  def test_fit_with_crop_should_not_overnudge
    bounds = [780, 520]
    assert_equal [375, 250], @x.fit_sizes_with_crop(bounds, :width => 260, :height => 250)
  end
end


image_proc_test.rb
require 'test/unit'
require 'fileutils'
require 'resize_test_helper'
require 'image_proc'

#class TestQuickProcessViaClassWithGeomString < Test::Unit::TestCase
#  def test_works
#    source = File.dirname(__FILE__) + '/input/horizontal.jpg'
#    dest = File.dirname(__FILE__) + '/output/resized.jpg'
#    assert_nothing_raised { ImageProc.resize(source, dest, "50x50") }
#    FileUtils.rm dest
#  end
#end

class TestQuickProcessWithOptions < Test::Unit::TestCase
  def test_resize_with_options
    source = File.dirname(__FILE__) + '/input/horizontal.jpg'
    dest = File.dirname(__FILE__) + '/output/resized.jpg'
    opts = {:height=>75, :fill=>true}
    begin
      assert_nothing_raised do
        path, w, h = ImageProc.resize_with_options(source, dest, opts)
      end
    ensure
      File.unlink(dest) rescue nil
    end
  end
  
  def test_raises_on_invalid_options
    assert_raise(ImageProc::InvalidOptions) do
      source = File.dirname(__FILE__) + '/input/horizontal.jpg'
      dest = File.dirname(__FILE__) + '/output/resized.jpg'
      opts = {:too => 4, :doo => 10}
      ImageProc.resize_with_options(source, dest, opts)
    end

    assert_raise(ImageProc::InvalidOptions) do
      source = File.dirname(__FILE__) + '/input/horizontal.jpg'
      dest = File.dirname(__FILE__) + '/output/resized.jpg'
      opts = {}
      ImageProc.resize_with_options(source, dest, opts)
    end
  end
end

class TestEngineAssignmentSticks < Test::Unit::TestCase
  def test_foreign_engine_assignment_sticks
    dummy = "foobar"
    ImageProc.engine = dummy
    assert_equal dummy, ImageProc.engine
    ImageProc.engine = nil
  end
end

if RUBY_PLATFORM =~ /darwin/i
  class TestImageProcSips < Test::Unit::TestCase
    def setup
      super
      @processor = ImageProcSips.new
      @landscapes.reject!{|e| e =~ /\.(png|gif)/}
      @portraits.reject!{|e| e =~ /\.(png|gif)/}
    end
  
    def test_sips_does_not_grok_pngs
      assert_raise(ImageProc::FormatUnsupported) do
        @processor.resize(INPUTS + '/horizontal.gif', OUTPUTS + '/horizontal.gif', "100x100")
      end
    end
  
    def test_sips_does_not_grok_gifs
      assert_raise(ImageProc::FormatUnsupported) do
        @processor.resize(INPUTS + '/horizontal.png', OUTPUTS + '/horizontal.png', "100x100")
      end
    end
  
    include ResizeTestHelper
  end
end

if(`which convert`)
  class TestImageProcConvert < Test::Unit::TestCase
    def setup
      super
      @processor = ImageProcConvert.new
    end
  
    include ResizeTestHelper
  end
end

begin
  require 'RMagick'
  class TestImageProcRmagick < Test::Unit::TestCase
    def setup
      super
      @processor = ImageProcRmagick.new
    end
  
    include ResizeTestHelper
  end
rescue LoadError
end

resize_test_helper.rb
module ResizeTestHelper
  INPUTS = File.expand_path(File.dirname(__FILE__) + '/input')
  OUTPUTS = File.expand_path(File.dirname(__FILE__) + '/output')
  
  def setup
    Dir.glob(File.dirname(__FILE__) + '/output/*.*').map{ |e| FileUtils.rm e }

    @extensions = ["jpg", "png", "gif"]
    @landscapes =  @extensions.map { | ext | "horizontal.#{ext}" }
    @portraits =  @extensions.map { | ext | "vertical.#{ext}" }
    @landscape_bounds = 780, 520
    @portraits_bounds = 466, 699
  end
  
  def test_get_bounds_raise_when_file_passed_does_not_exist
    assert_raise(ImageProc::Error) do
      @processor.get_bounds("/tmp/__non_existent/#{Time.now.to_i}")
    end
  end
  
  def test_get_bounds_raise_when_file_passed_is_not_an_image
    assert_raise(ImageProc::Error) do
      @processor.get_bounds(INPUTS + '/not.an.image.tmp')
    end
  end
  
  def test_properly_detects_bounds
    @landscapes.map do |file| 
      assert_equal @landscape_bounds, @processor.get_bounds(INPUTS + "/" + file)
    end
    
    @portraits.map do |file| 
      assert_equal @portraits_bounds, @processor.get_bounds(INPUTS + "/" + file)
    end
  end
  
  def test_resize_raises_when_trying_to_overwrite
    assert_raise(ImageProc::NoOverwrites) do
      @processor.resize INPUTS + '/' + @landscapes[0], INPUTS + '/' + @portraits[0], "100x100"
    end
  end
  
  def test_resize_raises_when_source_missing
    missing_input = "/tmp/___imageproc_missing.jpg"
    assert !File.exist?(missing_input)
    assert_raise(ImageProc::MissingInput) do
      @processor.resize missing_input, OUTPUTS + '/zeoutput.jpg', "100x100"
    end
  end
  
  def test_resize_raises_wheh_destination_dir_missing
    missing_output = "/tmp/___imageproc_missing/__missing.jpg"
    from = (INPUTS + '/' + @landscapes[0])
    
    assert !File.exist?(File.dirname(missing_output))
    assert_raise(ImageProc::NoDestinationDir) do
      @processor.resize from, missing_output, "100x100"
    end    
  end
  
  def test_resize_exact
    names = (@landscapes  + @portraits)
    sources = names.map{|file| INPUTS + "/" + file }

    sources.each_with_index do | source, index |
      assert_nothing_raised do
        path, w, h  = @processor.resize_exact(source, OUTPUTS + '/' + names[index], 65, 65)
        assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
      end
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [65, 65], get_bounds(result_p), "The image should have been resized exactly"
    end
  end
  
  def test_resize_fitting_proportionally_into_square
    with_each_horizontal_path_and_name do | source, name |
      assert_nothing_raised do
        path, w, h  = @processor.resize_fit(source, OUTPUTS + '/' + name, 300, 300)
        assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
      end
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [300, 200], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into rect proortionally"
    end
    
    with_each_vertical_path_and_name do | source, name |
      assert_nothing_raised { @processor.resize_fit(source, OUTPUTS + '/' + name, 300, 300) }
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [200, 300 ], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into rect proortionally"
    end
  end
  
  def test_fit_square_is_alias_for_proportional_resize
    with_each_horizontal_path_and_name do | source, name |
      assert_nothing_raised do
        path, w, h  = @processor.resize_fit_square(source, OUTPUTS + '/' + name, 300)
        assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
      end
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [300, 200], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into rect proortionally"
    end
  end
  
  def test_resize_fitting_proportionally_into_portrait
    with_each_horizontal_path_and_name do | source, name |
      assert_nothing_raised do
        path, w, h  = @processor.resize_fit(source, OUTPUTS + '/' + name, 20, 100)
        assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
      end

      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [20, 13], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into rect proortionally"
    end
  end
  
  def test_resize_fitting_proportionally_into_landscape
     with_each_vertical_path_and_name do | source, name |
       assert_nothing_raised do
         path, w, h  = @processor.resize_fit(source, OUTPUTS + '/' + name, 100, 20)
         assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
       end

       result_p = OUTPUTS + '/' + File.basename(source)
       assert File.exist?(result_p), "#{result_p} should have been created"
       assert_equal [13, 20], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into rect proortionally"
     end
  end
  
  def test_replaces_wildcards_in_filenames_after_resizing
    source = INPUTS + '/' + @landscapes[0]
    with_wildcards = OUTPUTS + '/resized_%dx%d' + File.extname(@landscapes[0])
    reference_path = with_wildcards % [300, 200]
    assert_nothing_raised do
      path, w, h  = @processor.resize_fit(source, with_wildcards, 300, 300)
      assert_equal path, reference_path, "The wildcards should be replaced with computed width and height and the file saved"
      assert_equal [300, 200], get_bounds(reference_path)
    end
  end
  
  def test_resize_is_alias_for_fit_with_geometry_string
     with_each_horizontal_path_and_name do | source, name |
       assert_nothing_raised { @processor.resize(source, OUTPUTS + '/' + name, "300x300") }
     
       result_p = OUTPUTS + '/' + File.basename(source)
       assert File.exist?(result_p), "#{result_p} should have been created"
       assert_equal [300, 200], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into rect proortionally"
     end
   
     with_each_vertical_path_and_name do | source, name |
       assert_nothing_raised do
          path = @processor.resize(source, OUTPUTS + '/' + name, "300x300")
          assert_kind_of String, path, "ImageProc#resize is legacy so it should return the path and nothing else"
          assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result"
       end
       
       result_p = OUTPUTS + '/' + File.basename(source)
       assert File.exist?(result_p), "#{result_p} should have been created"
       assert_equal [200, 300 ], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into rect proortionally"
     end
  end
  
  def test_resize_fit_width
    with_each_horizontal_path_and_name do | source, name |
      assert_nothing_raised do
         path, w, h = @processor.resize_fit_width(source, OUTPUTS + '/' + name, 400)
         assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
      end
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [400, 267], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been into width"
    end
    
    with_each_vertical_path_and_name do | source, name |
      assert_nothing_raised do
         path, w, h = @processor.resize_fit_width(source, OUTPUTS + '/' + name, 400)
      end
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [400, 600 ], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into width"
    end
  end
  
  def test_resize_fit_height
    with_each_horizontal_path_and_name do | source, name |
      assert_nothing_raised do
         path, w, h = @processor.resize_fit_height(source, OUTPUTS + '/' + name, 323)
         assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
      end
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [485, 323], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been into width"
    end
    
    with_each_vertical_path_and_name do | source, name |
      assert_nothing_raised do
         path, w, h = @processor.resize_fit_width(source, OUTPUTS + '/' + name, 323)
         assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
      end
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [323, 485], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been fit into width"
    end
  end
  
  def test_resize_fit_fill
    with_each_horizontal_path_and_name do | source, name |
      assert_nothing_raised do
         path, w, h = @processor.resize_fit_fill(source, OUTPUTS + '/' + name, 260, 250)
         assert_equal OUTPUTS + '/' + File.basename(source), path, "The proc should return the path to the result as first ret"
      end
      
      result_p = OUTPUTS + '/' + File.basename(source)
      assert File.exist?(result_p), "#{result_p} should have been created"
      assert_equal [375, 250], get_bounds(result_p), "The image of #{get_bounds(source).join("x")} should have been into width"
    end
  end
  
  private
    def get_bounds(of)
      if RUBY_PLATFORM =~ /darwin/i
        `sips #{of} -g pixelWidth -g pixelHeight`.scan(/(pixelWidth|pixelHeight): (\d+)/).to_a.map{|e| e[1]}
      else
        `identify #{of}`.scan(/(\d+)x(\d+)/)[0]
      end.map{|e| e.to_i}
    end
    
    def with_each_vertical_path_and_name
      @portraits.each do | file |
        yield(INPUTS + "/" + file, file)
      end
    end
    
    def with_each_horizontal_path_and_name
      @landscapes.each do | file |
        yield(INPUTS + "/" + file, file)
      end
    end
    
end